### PR TITLE
Preserve the callback behaviour of Meteor.subscribe for repeated subscriptions

### LINKED ISSUE #7: onReady callback isn't called when subscription is already active
When providing an onReady callback to `Meteor.subscribe`, it will always be called, even if an identical subscription is already active. `subsCache.subscribe` does not behave in this way; if a subscription with the same arguments already exists, the callback is ignored.

For example

```
var l = function() { console.log("onReady"); };
var sc = new SubsCache({
  expireAter: 5,
  cacheLimit: 10
});

Meteor.subscribe("mysub", l); // "onReady"
Meteor.subscribe("mysub", l); // "onReady"

sc.subscribe("mysub", l); // "onReady"
sc.subscribe("mysub", l); // nothing, never
```

### PR DESCRIPTION
**Pull request by BudgieInWA**
Any onReady callback that is provided to repeated .subscribe(args...) calls is passed on to existing
cachedSub.onReady.

For the other callbacks, a list of all callbacks is kept and the real subscription is given a delegation
function that delegates to the list.

The tests pass on Meteor 1.0.3.2 and 1.1.0.2.

This closes #7.

### CODE REVIEW

**Comment by BudgieInWA** [comment]
@ccorcos I'll justify here my reasons for all the changes.

Lines [61 - 75](https://github.com/ccorcos/meteor-subs-cache/pull/8/files#diff-5f33349ead0d4c1c61e55fe5a5694fd6R61) are good to have because they mimic the behaviour of `Meteor.subscribe`. That is, callbacks are not sent as subscription arguments, so a new subscription differing only in the callbacks should be considered the same.

Lines [81 - 82](https://github.com/ccorcos/meteor-subs-cache/pull/8/files#diff-5f33349ead0d4c1c61e55fe5a5694fd6R81) implement the behaviour that I think you were proposing.

Lines [83 - 86](https://github.com/ccorcos/meteor-subs-cache/pull/8/files#diff-5f33349ead0d4c1c61e55fe5a5694fd6R83) are needed to mimic the behaviour of `Meteor.subscribe` in how it handles onStop and onError callbacks. You can comment those lines out and run the callback tests to see the difference in behaviour.

The rest of the code is needed to make the previous lines work.

**Comment by ccorcos** [comment]
Hmm. You make a good point about eliminating the callback from the hash. 

Lets clean it up a little bit and break it down into smaller pieces. Maybe something like this:

```coffee
hasCallbacks = (argsList) ->
  if argsList.length > 0
    last = argsList[argsList.length-1]
    return _.isFunction(last) or _.isFunction(last?.onReady)
  return false

withoutCallbacks = (argsList) ->
  if hasCallbacks(argsList)
    return argsList[0...argsList.length-2]
  else
    return argsList

registerCallbacks: (argsList) ->
  if hasCallbacks(argsList)
    @callbacks = argsList[argsList.length-1]
  else
    @callbacks = null

handleOnReady: ->
  if _.isFunction(@callbacks) then @callbacks()

handleOnError: ->
  @callbacks?.onError?()

handleOnStop: ->
  @callbacks?.onStop?()
```

Then you can use these functions in the main code:

```coffee
  subscribeFor: (expireTime, args...) ->
    if Meteor.isServer
      # If we're using fast-render for SSR
      Meteor.subscribe.apply(Meteor.args)
    else
      @registerCallbacks(args)
      hash = EJSON.stringify(withoutCallbacks(args))
      cache = @cache
      if hash of cache
        # if we find this subscription in the cache, then restart it
        cache[hash].restart()
        @handleOnReady()
```

Breaking it down into semantic functions like this requires much less explanation, ya dig?

**Comment by BudgieInWA** [comment]
Good call on breaking it down into smaller functions. I have done a bit of a refactor.

File: src/subsCache.coffee [hunk 1]
```diff
@@ -1,6 +1,41 @@
 # debug = (args...) -> console.log.apply(console, args)
 debug = (args...) -> return
 
+makeDelegatingCallbackFn = (cachedSub, callbackName)->
```
Comment (ccorcos) [inline src/subsCache.coffee]: this lingo is confusing me. Delegating?

File: src/subsCache.coffee [hunk 2]
```diff
@@ -51,22 +86,24 @@ class @SubsCache
...
     else
-      hash = EJSON.stringify(args)
+      hash = EJSON.stringify(withoutCallbacks args)
       cache = @cache
       if hash of cache
-        # if we find this subscription in the cache, then restart it
+        # if we find this subscription in the cache, then rescue the callbacks
+        # and restart the cached subscription
+        if hasCallbacks args
+          cache[hash].registerCallbacks callbacksFromArgs args
```
Comment (ccorcos) [inline src/subsCache.coffee]: I think all we need to do is call onReady here. The onStop callback should still be hooked up from before.

**Comment by ccorcos** [comment]
Much better. Lets work on the lingo a little. Lets call these hooks. And lets put everything in the sub object itself. Something like this:

```coffee
  subscribeFor: (expireTime, args...) ->
    # {clip}
        # if we find this subscription in the cache, then restart it
        # and set the hooks from the arguments
        sub = cache[hash]
        sub.readyHook = 
        sub.errorHook = 
        sub.stopHook = 
        cache[hash].restart()
      else
        # {clip}
        cachedSub =
          readyHook:
          errorHook:
          stopHook:
          restart: ->
            # if we'are restarting, then stop the timer
            Meteor.clearTimeout(@timerId)
            @start()
            @readyHook?()
          stopNow: ->
            @sub.stop()
            @stopHook?()
            delete cache[@hash]
```

What do you think? I think I'm missing a couple places...

File: src/subsCache.coffee (hunk 1 above)
Comment (BudgieInWA) [inline src/subsCache.coffee]: I'm not completely happy with it either. I want to convey that calling this function causes the registered callbacks to be called in the same way. That is, "you can pretend that this is the original callback".

Maybe "makeOnStopPassthroughFn" or something along those lines? I don't know.

**Comment by BudgieInWA** [comment]
Good idea, I'll get to that soon.

**Comment by BudgieInWA** [comment]
Phew, I finally got some time for this.

**Comment by BudgieInWA** [comment]
Bump. 

@ccorcos, how do these changes look coming back with fresh eyes, as it were?

**Comment by talha-asad** [comment]
Should we get this documented?

**Comment by BudgieInWA** [comment]
My understanding is that the behaviour is now in line with the docs. I think

> `sub = subsCache.subscribe(...)` creates a subscription just like `Meteor.subscribe`

pretty much covers it, but if you think the callback behaviour deserves explicit mentioning, then I say go for it.

**Pull request merged**